### PR TITLE
test/msgr/cmake: link global-static to avoid DPDK EAL double initialization

### PR DESCRIPTION
**Pull request by rosinL**
The dpdk library initializes the EAL using constructors and global
variables, and cannot be re-initialized. Both test application and
libceph-common.so have a DPDK constructor that causes repeated
initialization. Link the ceph-common static library so that there
is only one constructor in test application.

Fixes: https://tracker.ceph.com/issues/42861
Signed-off-by: Chunsong Feng <fengchunsong@huawei.com>
Signed-off-by: luo rixin <luorixin@huawei.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by fengchunsong** [comment]
@tchaikov @adamemerson @cbodley   @yuyuyu101 Please help review, thank you！

**Comment by fengchunsong** [comment]
The new solution is to modify the test application to link the ceph-common static library,this PR will close

**Comment by fengchunsong** [comment]
@tchaikov please help to review,thanks

**Comment by tchaikov** [comment]
@fengchunsong will take a look shortly.

File: src/test/msgr/CMakeLists.txt [hunk 1]
```diff
@@ -16,15 +16,15 @@ add_executable(ceph_test_async_networkstack
   test_async_networkstack.cc
   $<TARGET_OBJECTS:unit-main>
   )
-target_link_libraries(ceph_test_async_networkstack global ${CRYPTO_LIBS} ${BLKID_LIBRARIES} ${CMAKE_DL_LIBS} ${UNITTEST_LIBS})
+target_link_libraries(ceph_test_async_networkstack global-static ${CRYPTO_LIBS} ${BLKID_LIBRARIES} ${CMAKE_DL_LIBS} ${UNITTEST_LIBS})
```
Comment (tchaikov) [inline src/test/msgr/CMakeLists.txt]: how is ceph_test_async_networkstack linked against two copies of DPDK?

File: src/test/msgr/CMakeLists.txt [hunk 2]
```diff
@@ -16,15 +16,15 @@ add_executable(ceph_test_async_networkstack
...
 
 #ceph_perf_msgr_server
 add_executable(ceph_perf_msgr_server perf_msgr_server.cc)
-target_link_libraries(ceph_perf_msgr_server os global ${UNITTEST_LIBS})
+target_link_libraries(ceph_perf_msgr_server os global-static ${UNITTEST_LIBS})
```
Comment (tchaikov) [inline src/test/msgr/CMakeLists.txt]: can we expose `dpdk::dpdk` as an `INTERFACE_LINK_LIBRARIES` exposed by os and ceph-common ?

File: src/test/msgr/CMakeLists.txt (hunk 1 above)
Comment (fengchunsong) [reply src/test/msgr/CMakeLists.txt]: target_link_libraries use the PUBLIC link property by default, Libraries following PUBLIC are linked to, and are made part of the link interface. check build/src/test/msgr/CMakeFiles/ceph_perf_msgr_client.dir/link.txt:
../../../lib/libceph-common.so ../../../lib/libjson_spirit.a ../../../lib/libcommon_    utf8.a ../../../lib/liberasure_code.a ../../../lib/libcrc32.a ../../../lib/libarch.a -lresolv ../../../boost/lib/libboost_thread.a ../../../boost/lib/libboost_chrono.a ../../../boo    st/lib/libboost_atomic.a ../../../boost/lib/libboost_system.a ../../../boost/lib/libboost_random.a ../../../boost/lib/libboost_program_options.a ../../../boost/lib/libboost_date_ti    me.a ../../../boost/lib/libboost_iostreams.a ../../../boost/lib/libboost_regex.a -lstdc++fs ../../../lib/libfmt.a /usr/lib/aarch64-linux-gnu/libblkid.so /usr/lib/aarch64-linux-gnu/    libcrypto.so /lib/aarch64-linux-gnu/libudev.so /usr/lib/aarch64-linux-gnu/libibverbs.so /usr/lib/aarch64-linux-gnu/librdmacm.so /usr/lib/aarch64-linux-gnu/libz.so ../../../lib/libc    ommon_async_dpdk.a -Wl,--whole-archive /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_bus_pci.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_cmdline.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_eal.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_ethdev.a /home/chunsong/ceph/obj-aarch64-linux-    gnu/src/dpdk/lib/librte_kvargs.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_mbuf.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_mempool.a /home/    chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_mempool_ring.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_pci.a /home/chunsong/ceph/obj-aarch64-linux-gnu    /src/dpdk/lib/librte_hash.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_net.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_pmd_i40e.a /home/chuns    ong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_pmd_ixgbe.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_pmd_hns3.a /home/chunsong/ceph/obj-aarch64-linux-gnu/sr    c/dpdk/lib/librte_pmd_hinic.a /home/chunsong/ceph/obj-aarch64-linux-gnu/src/dpdk/lib/librte_ring.a -Wl,--no-whole-archive  -Wl,-lnuma -Wl,-lpthread,-ldl ../../../lib/libgmock.a ../    ../../lib/libgtest.a -lpthread -ldl
it link libceph-common.so and librte_eal.a ,this cause two copies of eal

File: src/test/msgr/CMakeLists.txt (hunk 2 above)
Comment (fengchunsong) [reply src/test/msgr/CMakeLists.txt]: ceph-common uses common-msg-objs. When WITH_DPDK is defined, common-msg-objs links common_async_dpdk. Extracting the dpdk separately will cause the common_async_dpdk link to fail.

**Comment by fengchunsong** [comment]
@tchaikov @liu-chunmei please help to review, thks

**Comment by tchaikov** [comment]
unstable please

**Comment by fengchunsong** [comment]
Do not modify the link attributes, use other methods to solve, are there any good suggestions?

**Comment by fengchunsong** [comment]
> unstable please
This problem is caused by link properties,maybe caused by https://github.com/ceph/ceph/pull/23650/files "-Wl,--whole-archive $<JOIN:${DPDK_ARCHIVES}, > -Wl,--no-whole-archive")
The libc ommon_async_dpdk.a depends on the dpdk library librte_eal.a. All libceph-common.so and applications that use the libc ommon_async_dpdk.a are linked to librte_eal.a, if each application is linked to libceph-common.so, double librte_eal.a appears.

**Comment by fengchunsong** [comment]
I re-modified the thread exit code by referring to the RDMADispatcher.

**Comment by fengchunsong** [comment]
@ tchaikov please help to review,thanks

**Comment by fengchunsong** [comment]
jenkins test dashboard cephadm

**Pull request closed without merging**